File: exam-system-java/server/src/main/java/com/yyjy/exam/question/repository/QuestionsRepository.java
package com.yyjy.exam.question.repository;

import com.yyjy.exam.entity.question.dto.QuestionsPageView;
import com.yyjy.exam.entity.question.entity.QuestionType;
import com.yyjy.exam.entity.question.entity.Questions;
import com.yyjy.exam.entity.question.entity.QuestionsTable;
import org.babyfish.jimmer.spring.repository.JRepository;

import java.util.List;

public interface QuestionsRepository extends JRepository<Questions, Long> {
	
	boolean existsByTypeAndTitle(QuestionType type, String title);
	
	default List<QuestionsPageView> getLast(int size) {
		QuestionsTable t = QuestionsTable.$;
		return sql().createQuery(t)
				       .orderBy(t.id().desc())
				       .select(t.fetch(QuestionsPageView.class))
				       .limit(size)
				       .execute();
	}
	
	default List<QuestionsPageView> getLastNotIn(int size, List<Long> popularIds) {
		QuestionsTable t = QuestionsTable.$;
		return sql().createQuery(t)
				       .where(t.id().notIn(popularIds))
				       .select(t.fetch(QuestionsPageView.class))
				       .limit(size)
				       .execute();
	}
	
	default QuestionsPageView findQuestionsPageViewById(long id) {
		QuestionsTable t = QuestionsTable.$;
		return sql().createQuery(t)
				       .where(t.id().eq(id))
				       .select(t.fetch(QuestionsPageView.class))
				       .fetchFirstOrNull();
	}
	
	default List<Long> findAllIds() {
		QuestionsTable t = QuestionsTable.$;
		return sql().createQuery(t)
				       .select(t.id())
				       .execute();
	}
	
	default List<Long> findIdsByTypeAndCategoryIds(QuestionType type, List<Long> categoryIds) {
		QuestionsTable t = QuestionsTable.$;
		return sql().createQuery(t)
				       .where(t.type().eq(type))
				       .where(t.categoryId().in(categoryIds))
				       .select(t.id())
				       .execute();
	}
}

File: exam-system-server/server/src/main/kotlin/com/yyjy/repository/QuestionsRepository.kt
package com.yyjy.repository

import com.yyjy.models.entity.Questions
import com.yyjy.models.entity.dto.QuestionsPageView
import com.yyjy.models.entity.id
import com.yyjy.models.entity.title
import org.babyfish.jimmer.spring.repository.KRepository
import org.babyfish.jimmer.sql.kt.ast.expression.desc
import org.babyfish.jimmer.sql.kt.ast.expression.eq
import org.babyfish.jimmer.sql.kt.ast.expression.valueIn
import org.babyfish.jimmer.sql.kt.ast.expression.valueNotIn

/**
 * @author Nyxcirea
 * date 2026/4/1
 * description: TODO
 */
interface QuestionsRepository : KRepository<Questions, Long> {
    fun existsByCategoryIdAndTitle(categoryId: Long, title: String): Boolean
    fun existsByTypeAndTitle(type: String, title: String): Boolean
    fun getLast(size: Int): List<QuestionsPageView> {
        return sql.createQuery(Questions::class) {
            orderBy(table.id.desc())
            select(table.fetch(QuestionsPageView::class))
        }.limit(size).execute()
    }

    fun getLastNotIn(size: Int, popularIds: List<Long>): List<QuestionsPageView> {
        return sql.createQuery(Questions::class) {
            where(table.id valueNotIn popularIds)
            select(table.fetch(QuestionsPageView::class))
        }.limit(size).execute()
    }
    fun findQuestionsPageViewById(id: Long): QuestionsPageView? {
        return sql.createQuery(Questions::class) {
            where(table.id eq id)
            select(table.fetch(QuestionsPageView::class))
        }.fetchFirstOrNull()
    }

    fun findAllIds(): List<Long> {
        return sql.createQuery(Questions::class) {
            select(table.id)
        }.execute()
    }

    fun titleIsRepeat(map: List<String>) {
        sql.createQuery(Questions::class) {
            where(table.title valueIn map)
            select(table.title)
        }.execute()
    }
}
